File: app/domain/services/blackboard_service.py
"""Blackboard 领域服务（Phase 1）。

提供 topic 级 publish / subscribe / pull（带 cursor）。
"""

from __future__ import annotations

from app.common.utils import new_blackboard_entry_id, now_iso
from app.domain.models.blackboard import BlackboardEntry, TopicCursor
from app.storage.file.blackboard_store import BlackboardStore


class BlackboardService:
    """发布/订阅黑板条目。"""

    def __init__(self, store: BlackboardStore) -> None:
        self._store = store
        # 内存 cursor 表：{(agent_id, session_id, topic): last_read}
        self._cursors: dict[tuple[str, str, str], int] = {}

    def publish(
        self,
        session_id: str,
        topic: str,
        publisher_id: str,
        content: str,
    ) -> BlackboardEntry:
        """发布一条黑板条目（追加写）。"""
        entry = BlackboardEntry(
            id=new_blackboard_entry_id(),
            session_id=session_id,
            topic=topic,
            publisher_id=publisher_id,
            content=content,
            created_at=now_iso(),
        )
        self._store.append(session_id, topic, entry.to_dict())
        return entry

    def pull(self, session_id: str, topic: str, agent_id: str) -> list[BlackboardEntry]:
        """拉取自上次读取以来的增量条目，更新 cursor。"""
        key = (agent_id, session_id, topic)
        cursor = self._cursors.get(key, 0)
        raw_entries = self._store.read_since(session_id, topic, cursor)
        entries = [BlackboardEntry.from_dict(d) for d in raw_entries]
        self._cursors[key] = cursor + len(entries)
        return entries

    def subscribe(self, session_id: str, topic: str) -> list[BlackboardEntry]:
        """读取 topic 全部条目（无 cursor）。"""
        return [BlackboardEntry.from_dict(d) for d in self._store.read_all(session_id, topic)]

    def list_topics(self, session_id: str) -> list[str]:
        return self._store.list_topics(session_id)

File: miniagents-assembled/src/main/java/com/codex/miniagents/domain/service/BlackboardService.java
package com.codex.miniagents.domain.service;

import com.codex.miniagents.domain.model.blackboard.BlackboardEntry;
import com.codex.miniagents.domain.model.blackboard.TopicCursor;
import com.codex.miniagents.infrastructure.storage.repository.BlackboardRepository;
import com.codex.miniagents.utils.IdUtils;

import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class BlackboardService {
    private final BlackboardRepository repository;

    private final Map<String, TopicCursor> cursors = new ConcurrentHashMap<>();

    public BlackboardService(BlackboardRepository repository) {
        this.repository = repository;
    }

    public BlackboardEntry publish(String sessionId, String topic, String publisherId, String content) {
        BlackboardEntry entry = new BlackboardEntry();
        entry.setId(IdUtils.newBlackboardEntryId());
        entry.setSessionId(sessionId);
        entry.setTopic(topic);
        entry.setPublisherId(publisherId);
        entry.setContent(content);
        entry.setCreatedAt(Instant.now());
        repository.append(sessionId, topic, entry);
        return entry;
    }

    public List<BlackboardEntry> pull(String sessionId, String topic, String agentId) {
        String key = cursorKey(agentId, sessionId, topic);
        TopicCursor cursor = cursors.computeIfAbsent(key, ignored -> new TopicCursor(agentId, topic, 0));
        int lastRead = cursor.getLastRead();
        List<BlackboardEntry> entries = repository.readSince(sessionId, topic, lastRead);
        cursor.setLastRead(lastRead + entries.size());
        return entries;
    }

    public List<BlackboardEntry> subscribe(String sessionId, String topic) {
        return repository.readAll(sessionId, topic);
    }

    public List<String> listTopics(String sessionId) {
        return repository.listTopics(sessionId);
    }

    public void deleteSession(String sessionId) {
        repository.deleteSession(sessionId);
    }

    private String cursorKey(String agentId, String sessionId, String topic) {
        return agentId + "::" + sessionId + "::" + topic;
    }
}
